contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,95 @@
+package com.linkedin.parseq.batching;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.internal.ContextImpl;
+import com.linkedin.parseq.internal.PlanContext;
+import com.linkedin.parseq.promise.SettablePromise;
+import com.linkedin.parseq.trace.Relationship;
+import com.linkedin.parseq.trace.ShallowTraceBuilder;
+
+public abstract class BatchingStrategy<K, G> {
+
+  private final ConcurrentHashMap<Long, List<InternalBatchEntry<K, Object>>> batches =
+      new ConcurrentHashMap<>();
+
+  @SuppressWarnings("unchecked")
+  public <T> Task<T> batchable(final String desc, final Callable<BatchEntry<K, T>> callable) {
+    return Task.async("batchable: " + desc, ctx -> {
+      BatchEntry<K, T> batchEntry = callable.call();
+
+      Long planId = ctx.getPlanId();
+      List<InternalBatchEntry<K, Object>> list = batches.get(planId);
+      if (list == null) {
+        list = new ArrayList<>();
+        List<InternalBatchEntry<K, Object>> existing = batches.putIfAbsent(planId, list);
+        if (existing != null) {
+          list = existing;
+        }
+      }
```

Comment:
Yeah, that was actually my first implementation. That form however makes one additional allocation and I think (haven't tested it yet) JVM would not be able to optimize it away because it is being passed in putIfAbsent() method. I decided to trade one extra ArrayList allocation on every call for one extra ConcurrentHashMap lookup on first call for the plan.